Magento2/Sniffs/Legacy/WidgetXMLSniff.php

```diff
@@ -0,0 +1,114 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
+namespace Magento2\Sniffs\Legacy;
+
+use DOMDocument;
+use PHP_CodeSniffer\Files\File;
+use PHP_CodeSniffer\Sniffs\Sniff;
+
+/**
+ * Test for obsolete nodes/attributes in the module.xml
+ */
+class WidgetXMLSniff implements Sniff
+{
+    private const ERROR_CODE_OBSOLETE = 'FoundObsoleteNode';
+    private const ERROR_CODE_FACTORY = 'FoundFactory';
+    private const ERROR_CODE_XML = 'WrongXML';
+
+    /**
+     * @inheritdoc
+     */
+    public function register(): array
+    {
+        return [
+            T_INLINE_HTML
+        ];
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function process(File $phpcsFile, $stackPtr)
+    {
+        if ($stackPtr > 0) {
+            return;
+        }
+
+        $xml = simplexml_load_string($this->getFormattedXML($phpcsFile));
+        if ($xml === false) {
+            $phpcsFile->addError(
+                sprintf(
+                    "Couldn't parse contents of '%s', check that they are in valid XML format",
+                    $phpcsFile->getFilename(),
+                ),
+                1,
+                self::ERROR_CODE_XML
+            );
+        }
+
+        $foundElements = $xml->xpath('/widgets/*[@type]');
+
+        foreach ($foundElements as $element) {
+            if (property_exists($element->attributes(), 'type')) {
```

Comment:
I would rearrange this to avoid excessive nesting

```php
foreach ($foundElements as $element) {
        if (!property_exists($element->attributes(), 'type')) {
            continue;
        }
        $type = $element['type'];
        if (preg_match('/\//', $type)) {
            $phpcsFile->addError(
                "Factory name detected: {$type}.",
                dom_import_simplexml($element)->getLineNo() - 1,
                self::ERROR_CODE_FACTORY
            );
        }
    }
}